File: packages/git-core/src/parse/refs.ts
/**
 * `git for-each-ref` and reflog parsers.
 *
 * Both formats are built from *our own* format strings, which is what makes them
 * reliable: `for-each-ref` separates fields with `%00` and records with a newline,
 * and the reflog format separates the same way. Free text (a commit subject, a
 * stash message) can contain anything except NUL and newline, and both are
 * guaranteed by ref/reflog semantics to have neither — a reflog message with a
 * newline is stored by Git with the newline escaped, which the live tests check
 * rather than assume.
 *
 * `for-each-ref` fields, in order:
 * `refname`, `objectname`, `objecttype`, `symref`, `upstream`, `upstream:track`,
 * `HEAD` marker (`*` for the checked-out branch, a space otherwise), and the
 * peeled `*objectname` for annotated tags.
 */
import { GitOutputParseError } from "../ports.js";
import { splitOnByte } from "../bytes/nul-framing.js";
import { CORE_LIMITS } from "../bytes/limits.js";

export type RefObjectType = "commit" | "tag" | "tree" | "blob" | "unknown";

export interface RefRecord {
  readonly refName: string;
  readonly oid: string;
  readonly objectType: RefObjectType;
  /** The ref a symbolic ref points at, e.g. `refs/heads/main` for `HEAD`. */
  readonly symref: string | null;
  readonly upstream: string | null;
  readonly upstreamTrack: UpstreamTrack | null;
  readonly isHead: boolean;
  /** Peeled target of an annotated tag; null for every other object type. */
  readonly peeledOid: string | null;
}

export interface UpstreamTrack {
  readonly ahead: number;
  readonly behind: number;
  readonly gone: boolean;
}

const FOR_EACH_REF_FORMAT = "for-each-ref";
const REFLOG_FORMAT = "reflog";

function decodeAsciiLine(bytes: Uint8Array, format: string): string {
  let text = "";
  for (const byte of bytes) {
    if (byte > 0x7f) {
      throw new GitOutputParseError(
        format,
        "unexpected non-ASCII byte in an ASCII field",
      );
    }
    text += String.fromCharCode(byte);
  }
  return text;
}

/** Split a newline-terminated stream into records, refusing a truncated tail. */
function splitLines(bytes: Uint8Array, format: string): Uint8Array[] {
  const lines: Uint8Array[] = [];
  let start = 0;
  for (let index = 0; index < bytes.byteLength; index += 1) {
    if (bytes[index] === 0x0a) {
      lines.push(bytes.subarray(start, index));
      start = index + 1;
    }
  }
  if (start < bytes.byteLength) {
    throw new GitOutputParseError(
      format,
      `output ended without a newline (${bytes.byteLength - start} trailing bytes)`,
    );
  }
  return lines;
}

export function parseForEachRef(
  bytes: Uint8Array,
  options: { maxEntries?: number } = {},
): RefRecord[] {
  const maxEntries = options.maxEntries ?? CORE_LIMITS.refListMaxEntries;
  const records: RefRecord[] = [];
  for (const line of splitLines(bytes, FOR_EACH_REF_FORMAT)) {
    if (line.byteLength === 0) {
      continue;
    }
    if (records.length >= maxEntries) {
      throw new GitOutputParseError(
        FOR_EACH_REF_FORMAT,
        `ref list exceeded ${maxEntries} entries`,
      );
    }
    const fields = splitOnByte(line, 0x00);
    // The last field (the peeled object name) is empty for every non-annotated
    // ref. A trailing NUL means the final field is empty, and a splitter that
    // pushes on each separator never pushes a field after the last one — so seven
    // parsed fields plus a trailing separator is eight fields, not a short record.
    const complete =
      fields.length === 8 || (fields.length === 7 && endsWithNul(line));
    if (!complete) {
      throw new GitOutputParseError(
        FOR_EACH_REF_FORMAT,
        `expected 8 fields per ref but found ${fields.length}`,
      );
    }
    const [
      refName,
      oid,
      objectType,
      symref,
      upstream,
      track,
      headMarker,
      peeledField,
    ] = fields;
    const peeled = peeledField ?? new Uint8Array(0);
    if (refName === undefined || refName.byteLength === 0) {
      throw new GitOutputParseError(
        FOR_EACH_REF_FORMAT,
        "ref record has no refname",
      );
    }
    const oidText = decodeAsciiLine(
      oid ?? new Uint8Array(0),
      FOR_EACH_REF_FORMAT,
    );
    if (oidText.length > 0 && !/^[0-9a-f]{40}$|^[0-9a-f]{64}$/.test(oidText)) {
      throw new GitOutputParseError(
        FOR_EACH_REF_FORMAT,
        `unexpected object name '${oidText}'`,
      );
    }
    records.push({
      refName: decodeAsciiLine(refName, FOR_EACH_REF_FORMAT),
      oid: oidText,
      objectType: parseObjectType(
        decodeAsciiLine(objectType ?? new Uint8Array(0), FOR_EACH_REF_FORMAT),
      ),
      symref: emptyToNull(
        decodeAsciiLine(symref ?? new Uint8Array(0), FOR_EACH_REF_FORMAT),
      ),
      upstream: emptyToNull(
        decodeAsciiLine(upstream ?? new Uint8Array(0), FOR_EACH_REF_FORMAT),
      ),
      upstreamTrack: parseUpstreamTrack(
        emptyToNull(
          decodeAsciiLine(track ?? new Uint8Array(0), FOR_EACH_REF_FORMAT),
        ),
      ),
      isHead:
        decodeAsciiLine(
          headMarker ?? new Uint8Array(0),
          FOR_EACH_REF_FORMAT,
        ) === "*",
      peeledOid: emptyToNull(
        decodeAsciiLine(peeled ?? new Uint8Array(0), FOR_EACH_REF_FORMAT),
      ),
    });
  }
  return records;
}

/** True when the record ended with its NUL separator, i.e. the last field is empty. */
function endsWithNul(bytes: Uint8Array): boolean {
  return bytes.byteLength > 0 && bytes[bytes.byteLength - 1] === 0x00;
}

function parseObjectType(text: string): RefObjectType {
  switch (text) {
    case "commit":
    case "tag":
    case "tree":
    case "blob":
      return text;
    default:
      // A future object type must not fail a ref listing; it is reported as
      // unknown so a caller can refuse the operation that needed it.
      return "unknown";
  }
}

function emptyToNull(text: string): string | null {
  return text.length === 0 ? null : text;
}

/**
 * Parse `%(upstream:track)`, whose possible values are `[ahead 1]`,
 * `[behind 2]`, `[ahead 1, behind 2]`, `[gone]` and empty.
 */
export function parseUpstreamTrack(text: string | null): UpstreamTrack | null {
  if (text === null) {
    return null;
  }
  if (text === "[gone]") {
    return { ahead: 0, behind: 0, gone: true };
  }
  const match = /^\[(?:ahead (\d+))?(?:, )?(?:behind (\d+))?\]$/.exec(text);
  if (match === null) {
    throw new GitOutputParseError(
      FOR_EACH_REF_FORMAT,
      `unexpected upstream track value '${text}'`,
    );
  }
  const ahead = match[1] === undefined ? 0 : Number.parseInt(match[1], 10);
  const behind = match[2] === undefined ? 0 : Number.parseInt(match[2], 10);
  return { ahead, behind, gone: false };
}

export interface ReflogRecord {
  /** Locator as Git printed it, e.g. `stash@{0}` or `HEAD@{3}`. */
  readonly locator: string;
  readonly oid: string;
  /**
   * Reflog subject bytes, e.g. `On main: stash message`. Kept as bytes: a stash
   * message is user text and may be non-UTF-8, and decoding for display is the
   * host's job (see `TextCodec`).
   */
  readonly subjectBytes: Uint8Array;
  /** Committer timestamp of the entry, epoch seconds. */
  readonly timestamp: number;
}

/**
 * Parse reflog output produced with `--format=%gd%x00%H%x00%gs%x00%ct`.
 *
 * The locator is where an entry currently sits and it *moves* when entries are
 * added or dropped, which is why callers must pair it with the object name and
 * re-check before any write.
 */
export function parseReflog(
  bytes: Uint8Array,
  options: { maxEntries?: number } = {},
): ReflogRecord[] {
  const maxEntries = options.maxEntries ?? CORE_LIMITS.stashListMaxEntries;
  const records: ReflogRecord[] = [];
  for (const line of splitLines(bytes, REFLOG_FORMAT)) {
    if (line.byteLength === 0) {
      continue;
    }
    if (records.length >= maxEntries) {
      throw new GitOutputParseError(
        REFLOG_FORMAT,
        `reflog exceeded ${maxEntries} entries`,
      );
    }
    const fields = splitOnByte(line, 0x00);
    if (fields.length !== 4) {
      throw new GitOutputParseError(
        REFLOG_FORMAT,
        `expected 4 fields per reflog entry but found ${fields.length}`,
      );
    }
    const [locator, oid, subject, timestamp] = fields;
    if (locator === undefined || locator.byteLength === 0) {
      throw new GitOutputParseError(
        REFLOG_FORMAT,
        "reflog entry has no locator",
      );
    }
    const timestampText = decodeAsciiLine(
      timestamp ?? new Uint8Array(0),
      REFLOG_FORMAT,
    );
    if (!/^\d+$/.test(timestampText)) {
      throw new GitOutputParseError(
        REFLOG_FORMAT,
        `unexpected timestamp '${timestampText}'`,
      );
    }
    records.push({
      locator: decodeAsciiLine(locator, REFLOG_FORMAT),
      oid: decodeAsciiLine(oid ?? new Uint8Array(0), REFLOG_FORMAT),
      subjectBytes: subject ?? new Uint8Array(0),
      timestamp: Number.parseInt(timestampText, 10),
    });
  }
  return records;
}

File: crates/refyard-core/src/parse/refs.rs
//! `git for-each-ref` and `git reflog show`.
//!
//! `for-each-ref` emits one ref per line with `%00`-separated fields; the trailing
//! `%(*objectname)` is the peeled object of an annotated tag, which is what the UI
//! needs to show the commit behind a tag. The last field is NUL-terminated by the
//! format string, so a ref with fewer fields than expected is a parse error rather
//! than a silently shifted record.
//!
//! Ref names and object names are ASCII by construction, so they are decoded
//! strictly: a byte above 0x7f in a ref name means the output is not what we think it
//! is, and inventing a replacement character there would hide that.

use crate::bytes::{decode_ascii, split_nul_frames};
use crate::problem::CoreError;

const FOR_EACH_REF: &str = "for-each-ref";
const REFLOG: &str = "reflog show";

/// One ref, with the fields the refs panel renders.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RefRecord {
    pub ref_name: String,
    pub oid: String,
    /// `commit`, `tag`, `tree`, `blob` — or `unknown` when Git reports something this
    /// build has not seen.
    pub object_type: String,
    /// The ref this one points at, when it is symbolic.
    pub symref: Option<String>,
    /// The upstream ref name, when one is configured.
    pub upstream: Option<String>,
    /// `[ahead 1, behind 2]`, `[gone]`, or absent.
    pub upstream_track: Option<UpstreamTrack>,
    /// True for the branch HEAD points at.
    pub is_head: bool,
    /// The peeled object of an annotated tag.
    pub peeled_oid: Option<String>,
}

/// What `%(upstream:track)` said.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct UpstreamTrack {
    pub ahead: u64,
    pub behind: u64,
    /// The upstream ref no longer exists.
    pub gone: bool,
}

/// One reflog entry, used for the stash list.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReflogRecord {
    pub locator: String,
    pub oid: String,
    /// The subject, kept as bytes: a stash message is user text.
    pub subject: Vec<u8>,
    pub timestamp: i64,
}

/// Parses `for-each-ref` output produced with the eight-field format.
pub fn parse_for_each_ref(bytes: &[u8], max_entries: usize) -> Result<Vec<RefRecord>, CoreError> {
    let mut records: Vec<RefRecord> = Vec::new();
    for line in bytes.split(|byte| *byte == b'\n') {
        if line.is_empty() {
            continue;
        }
        if records.len() >= max_entries {
            return Err(CoreError::output_unparsable(
                FOR_EACH_REF,
                format!("ref listing exceeded {max_entries} entries"),
            ));
        }
        let normalized = normalize_line(line);
        let fields = split_nul_frames(&normalized, FOR_EACH_REF)?;
        // Eight fields are written; the format's trailing %00 makes the final
        // (peeled-object) field terminate the record, so a tag with no peeled object
        // arrives as an empty eighth field rather than a missing one.
        if fields.len() != 8 {
            return Err(CoreError::output_unparsable(
                FOR_EACH_REF,
                format!("expected 8 fields per ref but found {}", fields.len()),
            ));
        }
        let field = |index: usize| -> Result<String, CoreError> {
            decode_ascii(fields[index], FOR_EACH_REF)
        };
        let optional = |index: usize| -> Result<Option<String>, CoreError> {
            if fields[index].is_empty() {
                Ok(None)
            } else {
                Ok(Some(decode_ascii(fields[index], FOR_EACH_REF)?))
            }
        };
        let object_type = field(2)?;
        records.push(RefRecord {
            ref_name: field(0)?,
            oid: field(1)?,
            object_type: if object_type.is_empty() {
                "unknown".to_string()
            } else {
                object_type
            },
            symref: optional(3)?,
            upstream: optional(4)?,
            upstream_track: parse_upstream_track(&field(5)?),
            is_head: field(6)? == "*",
            peeled_oid: optional(7)?,
        });
    }
    Ok(records)
}

/// `%(upstream:track)` writes `[ahead 1]`, `[behind 2]`, `[ahead 1, behind 2]` or
/// `[gone]`; an empty field means the branch is level with its upstream.
pub fn parse_upstream_track(text: &str) -> Option<UpstreamTrack> {
    let inner = text.trim();
    if inner.is_empty() {
        return None;
    }
    let inner = inner.strip_prefix('[')?.strip_suffix(']')?;
    if inner == "gone" {
        return Some(UpstreamTrack {
            ahead: 0,
            behind: 0,
            gone: true,
        });
    }
    let mut ahead = 0u64;
    let mut behind = 0u64;
    for part in inner.split(',') {
        let part = part.trim();
        if let Some(value) = part.strip_prefix("ahead ") {
            ahead = value.trim().parse::<u64>().ok()?;
            continue;
        }
        if let Some(value) = part.strip_prefix("behind ") {
            behind = value.trim().parse::<u64>().ok()?;
            continue;
        }
        // An unrecognised track form is not something to guess about.
        return None;
    }
    Some(UpstreamTrack {
        ahead,
        behind,
        gone: false,
    })
}

/// Parses `reflog show --format=%gd%x00%H%x00%gs%x00%ct`.
pub fn parse_reflog(bytes: &[u8], max_entries: usize) -> Result<Vec<ReflogRecord>, CoreError> {
    let mut records: Vec<ReflogRecord> = Vec::new();
    for line in bytes.split(|byte| *byte == b'\n') {
        if line.is_empty() {
            continue;
        }
        if records.len() >= max_entries {
            return Err(CoreError::output_unparsable(
                REFLOG,
                format!("reflog exceeded {max_entries} entries"),
            ));
        }
        let normalized = normalize_line(line);
        let fields = split_nul_frames(&normalized, REFLOG)?;
        if fields.len() != 4 {
            return Err(CoreError::output_unparsable(
                REFLOG,
                format!(
                    "expected 4 fields per reflog entry but found {}",
                    fields.len()
                ),
            ));
        }
        let timestamp = decode_ascii(fields[3], REFLOG)?
            .trim()
            .parse::<i64>()
            .map_err(|_| CoreError::output_unparsable(REFLOG, "commit time is not a number"))?;
        records.push(ReflogRecord {
            locator: decode_ascii(fields[0], REFLOG)?,
            oid: decode_ascii(fields[1], REFLOG)?,
            // Not decoded: a stash message is user text and stays bytes until the
            // codec decides how to display it.
            subject: fields[2].to_vec(),
            timestamp,
        });
    }
    Ok(records)
}

/// Git separates records with `\n` and fields with NUL, so a record's own NUL run
/// ends at the newline. Appending a terminator lets the shared splitter be used on
/// one line without special-casing the last field.
fn normalize_line(line: &[u8]) -> Vec<u8> {
    let mut owned = line.to_vec();
    if owned.last() != Some(&0) {
        owned.push(0);
    }
    owned
}

#[cfg(test)]
mod tests {
    use super::*;

    /// One record as Git writes it: fields joined by NUL, the record *terminated* by
    /// a NUL, then the newline that separates records.
    fn line(fields: &[&str]) -> String {
        format!(
            "{}\u{0}\n",
            fields
                .iter()
                .map(|value| value.to_string())
                .collect::<Vec<_>>()
                .join("\u{0}")
        )
    }

    #[test]
    fn reads_a_branch_with_its_upstream_and_head_marker() {
        let text = line(&[
            "refs/heads/main",
            "1111111111111111111111111111111111111111",
            "commit",
            "",
            "refs/remotes/origin/main",
            "[ahead 2, behind 1]",
            "*",
            "",
        ]);
        let records = parse_for_each_ref(text.as_bytes(), 100).expect("parses");
        assert_eq!(records.len(), 1);
        let record = &records[0];
        assert_eq!(record.ref_name, "refs/heads/main");
        assert!(record.is_head);
        assert_eq!(record.upstream.as_deref(), Some("refs/remotes/origin/main"));
        assert_eq!(
            record.upstream_track,
            Some(UpstreamTrack {
                ahead: 2,
                behind: 1,
                gone: false
            })
        );
        assert_eq!(record.peeled_oid, None);
    }

    #[test]
    fn reads_the_peeled_object_of_an_annotated_tag() {
        // Without the peeled name the UI cannot show which commit a tag points at.
        let text = line(&[
            "refs/tags/v1",
            "2222222222222222222222222222222222222222",
            "tag",
            "",
            "",
            "",
            "",
            "3333333333333333333333333333333333333333",
        ]);
        let records = parse_for_each_ref(text.as_bytes(), 100).expect("parses");
        assert_eq!(
            records[0].peeled_oid.as_deref(),
            Some("3333333333333333333333333333333333333333")
        );
        assert_eq!(records[0].object_type, "tag");
    }

    #[test]
    fn reports_a_gone_upstream_as_gone_rather_than_as_level() {
        let track = parse_upstream_track("[gone]").expect("track");
        assert!(track.gone);
        assert_eq!(track.ahead, 0);
        assert_eq!(track.behind, 0);
    }

    #[test]
    fn treats_an_empty_track_field_as_no_upstream_status() {
        assert_eq!(parse_upstream_track(""), None);
    }

    #[test]
    fn refuses_an_unrecognised_track_form_instead_of_guessing() {
        assert_eq!(parse_upstream_track("[sideways 3]"), None);
    }

    #[test]
    fn refuses_a_ref_line_with_the_wrong_field_count() {
        // A shifted parse would attach another ref's object name to this name.
        let text = "refs/heads/main\u{0}1111111111111111111111111111111111111111\u{0}commit\n";
        assert!(parse_for_each_ref(text.as_bytes(), 100).is_err());
    }

    #[test]
    fn reads_reflog_entries_with_a_raw_subject() {
        let text = format!(
            "stash@{{0}}\u{0}{}\u{0}WIP on main: 1234567 base\u{0}1770000000\n",
            "4".repeat(40)
        );
        let records = parse_reflog(text.as_bytes(), 100).expect("parses");
        assert_eq!(records.len(), 1);
        assert_eq!(records[0].locator, "stash@{0}");
        assert_eq!(records[0].subject, b"WIP on main: 1234567 base");
        assert_eq!(records[0].timestamp, 1_770_000_000);
    }

    #[test]
    fn bounds_the_ref_listing() {
        let mut text = String::new();
        for index in 0..3 {
            text.push_str(&line(&[
                &format!("refs/heads/b{index}"),
                "1111111111111111111111111111111111111111",
                "commit",
                "",
                "",
                "",
                "",
                "",
            ]));
        }
        assert!(parse_for_each_ref(text.as_bytes(), 2).is_err());
        assert!(parse_for_each_ref(text.as_bytes(), 3).is_ok());
    }
}
